earth2studio/utils/coords.py

```diff
@@ -339,3 +339,93 @@ def split_coords(
     values = reduced_coords.pop(dim)
     xs = [xi.squeeze(dim_index) for xi in x.split(1, dim=dim_index)]
     return xs, reduced_coords, values
+
+
+def convert_multidim_to_singledim(
+    coords: CoordSystem, return_mapping: bool = False
+) -> CoordSystem:
+    """Converts a set of coordinates from a complex coordinate system, which has some
+    coordinates with multidimensional arrays, into a simple coordinate system
+    containing only one-dimensional arrays.
+
+    This conversion is done by creating individual indexes that are enumerations of
+    the complex coordinate.
+
+    Assumptions
+    -----------
+    This code assumes that if a coordinate entry is n-dimensional, then the following
+    (n-1) coordinate entries in the ordered dictionary have the same shape and represent
+    the same multidimensional grid.
+
+    Example
+    -------
+
+    Suppose we have lat/lon coordinates represented by 2-dimensional grids.
+    ```python
+    lat = np.linspace(0, 1, 10)
+    lon = np.linspace(0, 1, 20)
+    LON, LAT = np.meshgrid(lat, lon)
+    c = CoordSystem({"lat": LAT, "lon": LON})
+
+    c1 = convert_multidim_to_singledim(c)
+    ```
+
+    `c1` has 2 keys - `x1` and `x2`, with shapes `(10,)` and `(20,)` respectively.
+
+    Parameters
+    ----------
+    coords : CoordSystem
+        CoordSystem to convert.
+    return_mapping : bool, optional
+        Whether to return the mapping between the multidimensional entries and their
+        reduced entries in the converted coordinate system.
+        By default false.
+
+    Returns
+    -------
+    CoordSystem
+        Converted coordinate system where each coordinate is 1-dimensional.
+    """
+
+    adjusted_coords = {}
+    mapping: dict[str, list[str]] = {}
+
+    items = list(coords.items())
+    i = 0
+    dim_number = 0
+    while i < len(items):
+        item = items[i]
+        k, v = item
+        ndim = v.ndim
+        if v.ndim < 2:
+            adjusted_coords[k] = v
+            i += 1
+        else:
+            s = v.shape
+            mapping[k] = []
+
+            for j in range(ndim):
+                if i + j > len(items) - 1:
+                    raise ValueError(
+                        "Assumed that if an n-dimensional coordinate exists, "
+                        "then there will be exactly n coordinates with the same shape."
+                    )
+
+                k1, v1 = items[i + j]
+                if v1.shape != s:
+                    raise ValueError(
+                        "Assumed that if an n-dimensional coordinate exists, "
+                        "then there will be exactly n coordinates with the same shape."
+                    )
+
+                adjusted_coords["i" + k1] = np.arange(s[j])
+                mapping[k].append("i" + k1)
+                mapping[k1] = mapping[k]
+                dim_number += 1
```

Comment:
`dim_number` is unused and can be removed here, right?

Reply:
Good catch, thanks.